src/mpid/ch4/shm/xpmem/xpmem_control.c

```diff
@@ -120,11 +105,125 @@ int MPIDI_XPMEM_ctrl_send_lmt_req_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+    if (MPIDI_XPMEM_global.coop_dmsg_queue) {
+        MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t *slmt_rts_hdr;
+        dmessage = MPIDI_XPMEM_global.coop_dmsg_queue;
+        slmt_rts_hdr = &dmessage->ctrl_hdr.xpmem_slmt_rts;
+        if (!OPA_load_int(&MPIDI_XPMEM_global.coop_counter
+                          [MPIDI_XPMEM_global.local_rank * MPIDI_XPMEM_global.num_local +
+                           slmt_rts_hdr->src_lrank])) {
+            DL_DELETE(MPIDI_XPMEM_global.coop_dmsg_queue, dmessage);
+            mpi_errno = MPIDI_XPMEM_ctrl_do_send_lmt_rts_cb(&dmessage->ctrl_hdr);
+            MPIR_ERR_CHECK(mpi_errno);
+
+            MPL_free(dmessage);
+        }
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_CTRL_SEND_LMT_RECV_FIN_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_ctrl_send_lmt_rts_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_XPMEM_coop_dmsg_t *dmessage;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_rts_t *slmt_rts_hdr =
+        &((MPIDI_SHM_ctrl_hdr_t *) ctrl_hdr)->xpmem_slmt_rts;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_RTS_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_RTS_CB);
+
+    if (OPA_load_int(&MPIDI_XPMEM_global.coop_counter
+                     [MPIDI_XPMEM_global.local_rank * MPIDI_XPMEM_global.num_local +
+                      slmt_rts_hdr->src_lrank])) {
+        /* Previous copy is not done, need to delay current copy to next round. */
+        dmessage =
+            (MPIDI_XPMEM_coop_dmsg_t *) MPL_malloc(sizeof(MPIDI_XPMEM_coop_dmsg_t), MPL_MEM_OTHER);
+        MPIR_ERR_CHKANDJUMP(dmessage == NULL, mpi_errno, MPI_ERR_NO_MEM, "**nomem");
+
+        MPIR_Memcpy(&dmessage->ctrl_hdr, ctrl_hdr, sizeof(MPIDI_SHM_ctrl_hdr_t));
+        DL_APPEND(MPIDI_XPMEM_global.coop_dmsg_queue, dmessage);
+    } else {
+        mpi_errno = MPIDI_XPMEM_ctrl_do_send_lmt_rts_cb(ctrl_hdr);
+        MPIR_ERR_CHECK(mpi_errno);
     }
 
   fn_exit:
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_CTRL_SEND_LMT_REQ_CB);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_RTS_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_XPMEM_ctrl_send_lmt_cts_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *sreq;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_cts_t *slmt_cts_hdr = &ctrl_hdr->xpmem_slmt_cts;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_CTS_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_CTRL_SEND_LMT_CTS_CB);
+
+    /* Sender gets the CTS packet from receiver and need to perform copy */
+    sreq = (MPIR_Request *) slmt_cts_hdr->sreq_ptr;
+
+    /* Complete XPMEM receive */
+    mpi_errno =
+        MPIDI_XPMEM_handle_lmt_cts_recv(slmt_cts_hdr->dest_offset, slmt_cts_hdr->data_sz,
+                                        slmt_cts_hdr->rreq_ptr, slmt_cts_hdr->dest_lrank, sreq);
+    MPIR_ERR_CHECK(mpi_errno);
```

Comment:
It is weird to "complete a receive" by the sender. For a single message, the sender does "send" and the receiver does "receive". I think I have already commented on this. Can you please fix it?

Reply:
Do you want to change into `MPIDI_XPMEM_handle_lmt_cts_pkt`? Maybe it's more relevant to the task sender does

Reply:
I actually do not think the subroutine is necessary because no other function reuses it.

Reply:
ok I can delete it.